### DISCUSSION TITLE
Add negative approach rates to difficulty adjust mod "extend limits".

### DISCUSSION THREAD

**Discussion by Nyanaro** [ppy/osu#22898] [Ideas]
Currently when clicking the "Extended Limits" option when customizing the "Difficulty Adjust" mod, the maximum approach rate increases to 11 as it is with AR10+DT normally, but the minimum option is still 0, despite normally being able to go down to -5 with AR0+HT.

While I would be somewhat happy with the extend limits approach rate option being expanded to -5 to match AR11, I would really enjoy if it were possible to bring it down even further, down to -50 for example (or some other really low number). Currently there exists custom stable clients as well as McOsu which let you do this.

TL;DR
Let users decrease approach rates to far below 0 when using the "Extend Limits" option for the "Difficulty Adjust" mod. :)

**Comment by Rekunan** [comment]
somewhat related: #16732

**Comment by Bubbleman532** [comment]
Would be great to see how feasible this is, a lot of top reading players enjoy pushing these limits

**Reply by bdach** [reply]
The implementation - *within reasonable limits* - takes approximately around five seconds (i.e. anything that's not https://github.com/ppy/osu/discussions/16732 which is a performance nightmare). The biggest teeth pain there I'd say is that a setting going "negative" doesn't really _look_ all that well, and to someone unfamiliar with the concept may initially make little sense.

**Reply by Bubbleman532** [reply]
There's no other real terminology for it though, since if you want approach rate to correlate linearly from ar0 being equal to every note on screen at the same time, you would have to rework approach rates. I also don't think people will be using that slider that way unless they _are_ familiar with reading challenges

**Reply by bdach** [reply]
Yeah I'm aware.

Maybe if the tooltip showed the preempt time then it would be easier to digest. Something like: if extended limits are off, stay with the normal values, but if extended limits are enabled, negative values become available, and preempt time is appended to the tooltip text. Call it expert mode if you will I guess.

**Reply by isakvik** [reply]
Sounds fine. Ideally, like Nyanaro said, I'd like to be able to select AR -40 if I want; this would however take up a sizeable range in the slider, but you could just hide negative AR selection behind another button labelled "Extend AR selection to negative values" or something of the sort. Having to use the arrow keys to select an exact value isn't a problem either if the UI turns out to be too small to be precise.

**Reply by bdach** [reply]
@apollo-dw already mentioned to me off-hand in the pp dev discord server that the range of the "negative AR" would have to be considerably wide because below-zero ARs do not 'hit as strong" as above-10. I'm not sure what -40 time-wise is, so I'm not yet sure that'd be considered "reasonable", but I'm not sure we have to be beholden to extending the 0-5 AR line back to the negatives either. Maybe we can rescale that.

That said all of the above is mostly guesswork on my behalf; as far as I'm concerned, for this feature the buck stops with @peppy. And even if we agree to add it, I think it's much more of a "nice to have" task than the stuff we have booked for this year already, since this is a feature that doesn't exist in stable anyways and we should be aiming for parity first (unless I'm dead wrong?)

**Reply by Nyanaro** [reply]
It doesn't exist in stable, which is why reading players run custom, modded stable clients and play offline, or use McOsu in order to lower the approach rate. The best reading player in the world, Aery, has almost completely stopped playing the game online, and has almost exclusively played offline on his custom client for the last 2 years in order to play approach rates below -5. While it isn't "feature parity" with the stable client, this would be a highly appreciated feature by reading players and get many of us frequently playing on lazer.

**Reply by bdach** [reply]
Oh come on now. I didn't say we're _not_ doing it _at all_ yet. I said we may not do it _in the near future_. We've got to prioritise.

**Reply by Nyanaro** [reply]
I understand that, but you said it would take 5 seconds earlier in this thread.

**Reply by bdach** [reply]
Maybe not now that there's talk of -40 or whatever. I wasn't aware that the range was to be extended _that_ much. We'll need to check that that is "sane", that it will not cause performance problems, probably figure out how to rescale that so that it doesn't look ridiculous.

**Reply by Nyanaro** [reply]
If I could suggest, just a textbox where players can type the AR themselves over having an adjustable slider would be nice. You could call it "AR override", or something along those lines. This way you could avoid a confusing adjustable slider with too many values and also not be concerned with where the limit should be.

This is my most anticipated feature of the Lazer client, and while I do understand you have other priorities, this feature would get many reading players actively using Lazer, which would help with the path towards ranked play as there would be more informed users giving feedback. :pleading_face:

**Reply by bdach** [reply]
> also not be concerned with where the limit should be

We *must* be concerned with where the limit should be, because pushing the limit of negative AR makes more stuff appear on screen, and therefore is a _performance_ consideration. There is likely to be _significant_ performance degradation with very large values if we just allow anything.

**Reply by peppy** [reply]
Does this also mean there's going to be -OD, -CS, etc.? It feels very weird offering one but not the others, and setting arbitrary limits per difficulty slider. If not, I do wonder if this should not be its own mod, rather than difficulty adjust having out-of-bound limits.

Something like "Infinite Approach" which starts at AR0 and offers localised adjustment. Although I can imagine the next issue is that players would want to use it alongside Difficulty Adjust, and compatibility support between them will be a pain.

**Reply by Nyanaro** [reply]
While I'm only concerned with approach rate myself, I don't see why CS and OD shouldn't get access to negative values. (-OD is a bit silly, as this is a rhythm game after all, but who am I to judge)

I'm sure there are people who want to play with those as well.

**Reply by isakvik** [reply]
> There is likely to be significant performance degradation with very large values if we just allow anything.

It's a performance consideration in the same way you might want to support maps with 1000 sliders a second, but it's not really a case worth optimizing for, even if allowed. Anyone pushing the envelope on how many objects are rendered at once would probably be aware that they're doing just that, and would be okay with an FPS drop.

> Does this also mean there's going to be -OD, -CS, etc.? It feels very weird offering one but not the others

It's not the most elegant solution, but there's merit to AR specifically because it's so tied to players' perception of the game that challenging it becomes its own rabbit hole of adaptability and improvement, so it matches what people would actually use (like the reason why this thread exists in the first place).

Anyway, made a PR last night (#24736) adding this to Difficulty Adjust, allowing down to -10. It's not really pushing the limits of what the game can do at all (and it's the simplest viable implementation), but it will allow players to explore the reading space a bit more.

**Reply by bdach** [reply]
> in the same way you might want to support maps with 1000 sliders a second, but it's not really a case worth optimizing for, even if allowed

You'll say that, but we want the game to be able to handle as much as can be thrown at it. Many hours in the past were spent using centipede as basically a performance benchmark for lazer. There are several loved maps that are - to put it mildly - pushing envelopes in terms of how much stuff is on screen ([mania bar line spam](https://github.com/ppy/osu/issues/23412), catch banana spam) and we're inevitably going to get reports about those too as soon as we push lazer out to more users.

You can call it a point of pride if you will, but in practice, it often pans out that looking away from that sort of stuff results in user reports anyway (and rightly so...)

**Reply by isakvik** [reply]
Of course you want the game to be as smooth as possible for every case, since it allows for more flexibility in what you can actually put in the game, but performance as a reason not to implement excessively low ARs is more like an ideological point of "not allowing bad performance in the client" rather than letting the players figure out where the bar of expectation of reasonable performance should be.

But yeah, you do have much more experience dealing with community complaints than me. If allowing stupid ARs at all through this setting would invite a flood of reports from say, mobile players, then there's much more of a reason to optimize further before implementing it.
